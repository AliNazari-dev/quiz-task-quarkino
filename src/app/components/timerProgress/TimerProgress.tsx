"use client";

import { useEffect } from "react";

type TimerProps = {
  onTimeout: () => void;
  setTime: (time: number) => void;
  time: number;
};

export default function TimerProgress({ onTimeout, time, setTime }: TimerProps) {
  useEffect(() => {
    if (time < 5) {
      const timer = setInterval(() => {
        setTime(time + 1);
      }, 1000);
      return () => clearInterval(timer);
    } else {
      onTimeout();
      setTime(0);
    }
  }, [time]);

  const progress = (time / 100) * 100;

  return (
    <div className='w-[300px] bg-green-700 h-3 rounded-lg mb-8'>
      <div
        className='bg-red-600 h-3 rounded-lg transition-all'
        style={{ width: `${progress * 20}%` }}></div>
      <div className='text-center mt-2 font-bold'>lost time : {time} s</div>
    </div>
  );
}
