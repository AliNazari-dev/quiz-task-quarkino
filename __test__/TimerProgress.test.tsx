// TimerProgress.test.tsx
import React from "react";
import { render, act, fireEvent, waitFor } from "@testing-library/react";
import TimerProgress from "../src/app/components/timerProgress/TimerProgress";

jest.useFakeTimers();

describe("TimerProgress", () => {
  const wrapperClassName = "timer-progress";
  // it("increments time and calls onTimeout when time reaches 5 seconds", async () => {
  //   const onTimeoutMock = jest.fn();
  //   const setTimeMock = jest.fn();

  //   const { getByText } = render(
  //     <TimerProgress onTimeout={onTimeoutMock} time={5} setTime={setTimeMock} />
  //   );

  //   // Ensure the initial UI is rendered correctly
  //   expect(getByText("lost time : 0 s")).toBeInTheDocument();

  //   // Advance time by 1 second
  //   act(() => {
  //     jest.advanceTimersByTime(1000);
  //   });

  //   // Ensure the time has been incremented
  //   expect(getByText("lost time : 1 s")).toBeInTheDocument();

  //   // Advance time by 4 more seconds
  //   act(() => {
  //     jest.advanceTimersByTime(4000);
  //   });

  //   // Ensure the time has been incremented to 5 seconds
  //   expect(getByText("lost time : 5 s")).toBeInTheDocument();

  //   // Ensure onTimeout is called
  //   await waitFor(() => {
  //     expect(onTimeoutMock).toHaveBeenCalledTimes(1);
  //   });

  //   // Ensure setTime is called with 0 after onTimeout
  //   await waitFor(() => {
  //     expect(setTimeMock).toHaveBeenCalledWith(0);
  //   });
  // });

  // it("clears interval on unmount", () => {
  //   const onTimeoutMock = jest.fn();
  //   const setTimeMock = jest.fn();

  //   const { unmount } = render(
  //     <TimerProgress onTimeout={onTimeoutMock} time={0} setTime={setTimeMock} />
  //   );

  //   // Unmount the component
  //   unmount();

  //   // Ensure clearInterval is called
  //   expect(clearInterval).toHaveBeenCalled();
  // });
  it("render correctly wrapper", () => {
    const onTimeoutMock = jest.fn();
    const setTimeMock = jest.fn();

    const { container } = render(
      <TimerProgress onTimeout={onTimeoutMock} time={2} setTime={setTimeMock} />
    );

    expect(container.firstChild).toHaveClass(wrapperClassName);
  });

  it("match snapshot", () => {
    const onTimeoutMock = jest.fn();
    const setTimeMock = jest.fn();

    const { container } = render(
      <TimerProgress onTimeout={onTimeoutMock} time={2} setTime={setTimeMock} />
    );

    expect(container).toMatchSnapshot();
  });
});
