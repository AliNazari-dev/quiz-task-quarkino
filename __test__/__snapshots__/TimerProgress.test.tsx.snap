// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimerProgress match snapshot 1`] = `
<div>
  <div
    class="w-[300px] bg-green-700 h-3 rounded-lg mb-8"
  >
    <div
      class="bg-red-600 h-3 rounded-lg transition-all"
      style="width: 40%;"
    />
    <div
      class="text-center mt-2 font-bold"
    >
      lost time : 
      2
       s
    </div>
  </div>
</div>
`;
